Automattic/wp-calypso
# Full Site Editing: render template parts on all site's pages

#### Changes proposed in this Pull Request

paAmJe-AI-p2

This removes our page hijacking and delegates header and footer override
functionality to themes, which allows us to override those section for
all of site's pages, not just static ones.

It also removes template data population from the plugin, and delegates
that responsibility to themes. The added benefit of this is that this data
population will work seamlessly on Dotcom too, avoiding the need to rely
on Headstart to provide it during site creation.

#### Testing instructions

1. This PR depends on the following Modern Business update: https://github.com/Automattic/themes/pull/1062
2. Assuming you'll be using local Gutenberg Docker dev environment, you'll need to map both the plugin and the theme folder in volumes (replace with your actual paths to Calypso and Themes repo):
      - `.../wp-calypso/apps/full-site-editing/full-site-editing-plugin:/var/www/html/wp-content/plugins/full-site-editing-plugin`
      - `.../themes/modern-business:/var/www/html/wp-content/themes/modern-business`
3. Restart your Docker containers for the above changes to take effect.
4. Delete all of the previously created Template and Template Parts CPT posts.
5. Activate Modern Business theme and verify that it has populated the template data correctly.
6. Check that everything works as expected while rending and editing some sample page.
7. Verify that header and footer are also rendered on dynamic pages.
8. Switch to Twenty Nineteen.
9. Since this theme doesn't support FSE, you should see the default header and footer, and no ability to edit them in the Editor.
10. Switch back to Modern Business.
11. Disable Full Site Editing plugin.
12. Verify that the theme still works fine without it. It should fall back to Twenty Nineteen defaults, and FSE functionality will no longer be available while editing.

Effectively, the themes now declare FSE support by inserting the appropriate FSE template data. If everything works out correctly, you should see the following entries for Template and Template Parts.

<img width="754" alt="Screenshot 2019-07-14 at 12 17 27" src="https://user-images.githubusercontent.com/1182160/61182302-85359380-a631-11e9-84be-badfedadf85a.png">
<img width="754" alt="Screenshot 2019-07-14 at 12 17 50" src="https://user-images.githubusercontent.com/1182160/61182303-85359380-a631-11e9-9395-169be1d8f23f.png">

One thing that has changed compared to previous version is that we are now saving theme specific taxonomy terms. This enables us to have template entries for multiple themes, and have the correct ones pulled in for currently active theme. For example, we could have `elegant-business-header` and `elegant-business-footer` alongside the ones listed above.

This also makes the theme switching flow a little more predictable. Let's say the user activates Modern Business and makes some custom modifications to its header that fit just this theme. After that, they switch to Elegant Business (similarly enhanced for FSE) and it will be provisioned with its default header. I think this is important because old theme's header (with potential custom user modifications) will likely not be suitable structurally or visually for a different theme. If they decide to revert back to Modern Business, they'll get its header again with any custom modifications that they have previously made for it. This seems like a better experience overall than just reverting to Modern Business default header (although we could easily implement that too).

Fixes https://github.com/Automattic/wp-calypso/issues/34560
